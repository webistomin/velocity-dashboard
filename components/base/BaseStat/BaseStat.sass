+b('base-stat')
  +e('measure')
    display: none

    +media($screen-sm)
      display: inline
      padding-left: rem(6px)
      font: var(--text-heading-3)

  +e('heading')
    margin-bottom: rem(4px)

    +media($screen-sm)
      display: flex
      align-items: baseline

  +e('percent')
    padding-left: rem(6px)

  +e('dynamic')
    +m('positive')
      color: var(--color-green)

    +m('negative')
      color: var(--color-red)
