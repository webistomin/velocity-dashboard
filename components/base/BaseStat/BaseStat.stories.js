import { number, text } from '@storybook/addon-knobs';
import BaseStat from './BaseStat';

export default {
  title: 'BaseStat',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
      previewTabs: {
        actions: {
          hidden: true,
        },
      },
    },
  },
};

export const Default = () => ({
  components: { BaseStat },
  props: {
    value: {
      type: Number,
      default: number('value', 50),
      required: true,
    },
    prevValue: {
      type: Number,
      default: number('prevValue', 49),
    },
    measure: {
      type: String,
      default: text('measure', 'km'),
    },
    icon: {
      type: String,
      default: text('icon', 'icon-lightning'),
    },
    color: {
      type: String,
      default: text('color', 'blue'),
    },
    align: {
      type: String,
      default: text('align', 'row'),
    },
  },
  template: `<base-stat
              :value='this.value'
              :prevValue='this.prevValue'
              :measure='this.measure'
              :icon='this.icon'
              :color='this.color'
              :align='this.align'
            />
  `,
});
