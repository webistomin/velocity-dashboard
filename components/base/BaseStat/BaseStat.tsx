import { VueComponent } from 'types/vue-components';
import { Component, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import BaseTitle from 'components/base/BaseTitle';
import { BaseIcon } from 'components/base/BaseIcon/BaseIcon';
import { animateValue } from '~/utils/animate-value';

import './BaseStat.sass';

export enum DynamicTypes {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
}

export type IBaseStatAlignTypes = 'row' | 'col';

export interface IBaseStatProps {
  value: number;
  prevValue?: number;
  measure: string;
  icon?: string;
  color?: string;
  align?: IBaseStatAlignTypes;
}

@Component({
  name: 'BaseStat',
})
export default class BaseStat extends VueComponent<IBaseStatProps> {
  animatedValue: IBaseStatProps['value'] = 0;

  @Prop()
  private readonly value!: IBaseStatProps['value'];

  @Prop({ default: 0 })
  private readonly prevValue: IBaseStatProps['prevValue'];

  @Prop()
  private readonly measure!: IBaseStatProps['measure'];

  @Prop()
  private readonly icon!: IBaseStatProps['icon'];

  @Prop()
  private readonly color!: IBaseStatProps['color'];

  @Prop({ default: 'col' })
  private readonly align!: IBaseStatProps['align'];

  mounted(): void {
    animateValue.call(this, 'animatedValue', 0, this.value, 2000);
  }

  get getStat(): string {
    // @ts-ignore
    return (((this.value - this.prevValue) / this.prevValue) * 100).toFixed(1);
  }

  get getDynamic(): string {
    const value = Number(this.getStat);
    if (value <= 0) {
      return DynamicTypes.NEGATIVE;
    }

    return DynamicTypes.POSITIVE;
  }

  get getIconName(): string {
    const dynamic = this.getDynamic;
    if (dynamic === DynamicTypes.NEGATIVE) {
      return 'icon-arrow-down';
    }

    return 'icon-arrow-up';
  }

  render(): VNode {
    return (
      <div class={`base-stat base-stat_align_${this.align}`}>
        {this.icon ? <BaseIcon class='base-stat__icon' name={this.icon} color={this.color} /> : null}
        <div class='base-stat__holder'>
          <div class='base-stat__heading'>
            <BaseTitle level={2} class='base-stat__title'>
              {this.animatedValue}
            </BaseTitle>
            <strong class='base-stat__measure'>{this.measure}</strong>
          </div>
          {this.prevValue ? (
            <span
              class={`base-stat__dynamic ${
                this.getDynamic === DynamicTypes.POSITIVE
                  ? 'base-stat__dynamic_positive'
                  : 'base-stat__dynamic_negative'
              }`}>
              <svg-icon name={this.getIconName} width={10} height={10} />
              <span class='base-stat__percent'>{this.getStat}%</span>
            </span>
          ) : null}
        </div>
      </div>
    );
  }
}
