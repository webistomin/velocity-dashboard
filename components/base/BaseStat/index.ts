import BaseStat from './BaseStat';

export default BaseStat;
