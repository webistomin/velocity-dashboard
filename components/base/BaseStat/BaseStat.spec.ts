import { mount } from '@vue/test-utils';
import BaseStat, { DynamicTypes } from './BaseStat';

describe('BaseStat', () => {
  const iconSelector = '.base-stat__icon';
  const measureSelector = '.base-stat__measure';
  const dynamicSelector = '.base-stat__dynamic';

  const factory = (propsData = {}, options = {}) => {
    return mount(BaseStat, {
      propsData: {
        value: 50,
        ...propsData,
      },
      stubs: ['svg-icon'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Return "Negative" if dynamic is negative', () => {
    const wrapper = factory({
      value: 20,
      prevValue: 40,
    });
    const result = wrapper.vm.getDynamic;
    const expected = DynamicTypes.NEGATIVE;
    expect(result).toEqual(expected);
  });

  it('Hide icon if props are not provided', () => {
    const wrapper = factory();
    const icon = wrapper.find(iconSelector);
    expect(icon.exists()).toBe(false);
  });

  it('Show icon if title prop is provided', () => {
    const wrapper = factory({
      icon: 'icon-bubble',
    });
    const icon = wrapper.find(iconSelector);
    expect(icon.exists()).toBe(true);
  });

  it('Hide measure if props are not provided', () => {
    const wrapper = factory();
    const measure = wrapper.find(measureSelector);
    expect(measure.exists()).toBe(false);
  });

  it('Show measure if title prop is provided', () => {
    const wrapper = factory({
      measure: 'km',
    });
    const measure = wrapper.find(measureSelector);
    expect(measure.exists()).toBe(true);
  });

  it('Set positive class if dynamic is positive', () => {
    const expectedResult = ['base-stat__dynamic_positive'];
    const wrapper = factory({
      value: 40,
      prevValue: 20,
    });
    const dynamic = wrapper.find(dynamicSelector);
    const classes = dynamic.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });
});
