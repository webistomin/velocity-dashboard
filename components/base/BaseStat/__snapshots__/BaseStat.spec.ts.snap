// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseStat Matches snapshot 1`] = `
"<div class=\\"base-stat base-stat_align_col\\">
  <div class=\\"base-stat__holder\\">
    <div class=\\"base-stat__heading\\">
      <h2 class=\\"title base-title base-title_level-2  base-stat__title\\">0</h2>
    </div>
  </div>
</div>"
`;
